api/v1/forkchoice.go

```diff
@@ -0,0 +1,234 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	"github.com/pkg/errors"
+)
+
+// ForkChoice is the data regarding the node's current fork choice context.
+type ForkChoice struct {
+	// JustifiedCheckpoint is the current justified checkpoint.
+	JustifiedCheckpoint phase0.Checkpoint
+	// FInalizedCheckpoint is the current finalized checkpoint.
+	FinalizedCheckpoint phase0.Checkpoint
+	// ForkChoiceNodes contains the fork choice nodes.
+	ForkChoiceNodes []*ForkChoiceNode
+}
+
+// MarshalJSON implements json.Marshaler.
+func (f *ForkChoice) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&forkChoiceJSON{
+		JustifiedCheckpoint: &f.JustifiedCheckpoint,
+		FinalizedCheckpoint: &f.FinalizedCheckpoint,
+		ForkChoiceNodes:     f.ForkChoiceNodes,
+	})
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+func (f *ForkChoice) UnmarshalJSON(input []byte) error {
+	var err error
+
+	var forkChoiceJSON forkChoiceJSON
+	if err = json.Unmarshal(input, &forkChoiceJSON); err != nil {
+		return errors.Wrap(err, "invalid JSON")
+	}
+
+	if forkChoiceJSON.JustifiedCheckpoint == nil {
+		return errors.New("justified checkpoint missing")
+	}
+	f.JustifiedCheckpoint = *forkChoiceJSON.JustifiedCheckpoint
+
+	if forkChoiceJSON.FinalizedCheckpoint == nil {
+		return errors.New("finalized checkpoint missing")
+	}
+	f.FinalizedCheckpoint = *forkChoiceJSON.FinalizedCheckpoint
+
+	if forkChoiceJSON.ForkChoiceNodes == nil {
+		return errors.New("fork choice nodes missing")
+	}
+	f.ForkChoiceNodes = forkChoiceJSON.ForkChoiceNodes
+
+	return nil
+}
+
+// String returns a string version of the structure.
+func (f *ForkChoice) String() string {
+	data, err := json.Marshal(f)
+	if err != nil {
+		return fmt.Sprintf("ERR: %v", err)
+	}
+	return string(data)
+}
+
+// forkChoiceJSON is the json representation of the struct.
+type forkChoiceJSON struct {
+	JustifiedCheckpoint *phase0.Checkpoint `json:"justified_checkpoint"`
+	FinalizedCheckpoint *phase0.Checkpoint `json:"finalized_checkpoint"`
+	ForkChoiceNodes     []*ForkChoiceNode  `json:"fork_choice_nodes"`
+}
+
+type ForkChoiceNodeValidity string
+
+const (
+	ForkChoiceNodeValidityInvalid    ForkChoiceNodeValidity = "INVALID"
+	ForkChoiceNodeValidityValid      ForkChoiceNodeValidity = "VALID"
+	ForkChoiceNodeValidityOptimistic ForkChoiceNodeValidity = "OPTIMISTIC"
+)
+
+func parseForkChoiceValidity(input string) (ForkChoiceNodeValidity, error) {
+	switch input {
+	case "INVALID":
+		return ForkChoiceNodeValidityInvalid, nil
+	case "VALID":
+		return ForkChoiceNodeValidityValid, nil
+	case "OPTIMISTIC":
+		return ForkChoiceNodeValidityOptimistic, nil
+	default:
+		return "", errors.New(fmt.Sprintf("invalid value for fork choice node validity: %s", input))
+	}
+}
```

Comment:
This should ideally bit its own enum, similar to that in `spec/dataversion.go` with the individual types being uint64 and having the same functions.

Reply:
Have implemented that now @mcdee 🙏